pyro/infer/advi.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.distributions import constraints, transform_to
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.poutine.util import prune_subsample_sites
+
+
+def _product(shape):
+    """
+    Computes the product of the dimensions of a given shape tensor
+    """
+    result = 1
+    for size in shape:
+        result *= size
+    return result
+
+
+class ADVI(object):
+    """
+    Base class for implementations of Automatic Differentiation Variational Inference.
+
+    Each derived class implements its own :meth:`sample_latent` method.
+
+    Assumes model structure and latent dimension are fixed, and all latent
+    variables are continuous.
+
+    :param callable model: a Pyro model
+    """
+    def __init__(self, model):
+        self.prototype_trace = None
+        self.base_model = model
+
+    def _setup_prototype(self, *args, **kwargs):
+        # run the model so we can inspect its structure
+        self.prototype_trace = poutine.block(poutine.trace(self.base_model).get_trace)(*args, **kwargs)
+        self.prototype_trace = prune_subsample_sites(self.prototype_trace)
+        self.latent_dim = sum(site["value"].view(-1).size(0)
+                              for site in self.prototype_trace.nodes.values()
+                              if site["type"] == "sample" and not site["is_observed"])
+
+    def sample_latent(self, *args, **kwargs):
+        """
+        Samples an encoded latent given the same ``*args, **kwargs`` as the
+        base ``model``.
+        """
+        raise NotImplementedError
+
+    def guide(self, *args, **kwargs):
+        """
+        An automatic guide with the same ``*args, **kwargs`` as the base ``model``.
+        """
+        # if we've never run the model before, do so now so we can inspect the model structure
+        if self.prototype_trace is None:
+            self._setup_prototype(*args, **kwargs)
+
+        latent = self.sample_latent(*args, **kwargs)
+        pos = 0
+        for name, site in self.prototype_trace.nodes.items():
+            if site["type"] == "sample" and not site["is_observed"]:
+                shape = site["fn"].shape()
+                size = _product(shape)
+                unconstrained_value = latent[pos:pos + size].view(shape)
+                pos += size
+                value = transform_to(site["fn"].support)(unconstrained_value)
+                pyro.sample(name, dist.Delta(value))
+        assert pos == len(latent)
+
+    def model(self, *args, **kwargs):
+        """
+        A wrapped model with the same ``*args, **kwargs`` as the base ``model``.
+        """
+        # wrap sample statement with a 0.0 poutine.scale to zero out unwanted score
+        with poutine.scale("advi_scope", 0.0):
+            self.sample_latent(*args, **kwargs)
```

Comment:
Since `self.model` and `self.guide` don't share any state, it seems like the right solution here would be to allow auxiliary guide variables in SVI?  That way `ADVI` could just return a guide.

That can be left for a potential followup PR, though.